### PR TITLE
Attach_disk&Detach_disk support, update persistent disk information in config driver accordingly

### PR DESCRIPTION
**Pull request by jianqiu**
This PR is including a mount of changes in attach_disk and detach_disk. In order to saving persistent disk information into config driver,  a new method `GetUserMetadataOnVirtualGuest ` created in `softlayer_helper` to retrieve user meta from virtual guest, then append/remove persistent disk information from it.

`+GetUserMetadataOnVirtualGuest` in `softlayer_helper`:
```go
func GetUserMetadataOnVirtualGuest(softLayerClient sl.Client, virtualGuestId int) ([]byte, error) {
	virtualGuestService, err := softLayerClient.GetSoftLayer_Virtual_Guest_Service()
	if err != nil {
		return []byte{}, bosherr.WrapError(err, "Creating VirtualGuestService from SoftLayer client")
	}

	attributes, err := virtualGuestService.GetUserData(virtualGuestId)
	if err != nil {
		return []byte{}, bosherr.WrapError(err, fmt.Sprintf("Getting metadata on VirtualGuest `%d`", virtualGuestId))
	}

	if len(attributes) == 0 {
		return []byte{}, bosherr.WrapError(err, fmt.Sprintf("Failed to get metadata on VirtualGuest `%d`", virtualGuestId))
	}

	sEnc := attributes[0].Value
	sDec, err := base64.StdEncoding.DecodeString(sEnc)
	if err != nil {
		return []byte{}, bosherr.WrapError(err, fmt.Sprintf("Failed to decode metadata returned from virtualGuest `%d`", virtualGuestId))
	}

	return sDec, nil
}
```

### CODE REVIEW

File: softlayer/vm/softlayer_vm.go [hunk 1]
```diff
@@ -144,11 +174,39 @@ func (vm SoftLayerVM) DetachDisk(disk bslcdisk.Disk) error {
...
+	if err != nil {
+		return bosherr.WrapError(err, fmt.Sprintf("Failed to revoke access of disk `%d` from virtual gusest `%d`", disk.ID(), virtualGuest.Id))
+	}
+
 	err = vm.detachVolumeBasedOnShellScript(virtualGuest, volume)
 	if err != nil {
 		return bosherr.WrapErrorf(err, "Failed to detach volume with id %d from virtual guest with id: %d.", volume.Id, virtualGuest.Id)
 	}
 
+	metadata, err := bslcommon.GetUserMetadataOnVirtualGuest(vm.softLayerClient, virtualGuest.Id)
```
Comment (maximilien) [inline softlayer/vm/softlayer_vm.go]: Why are you not checking `if err != nil` here?

File: softlayer/vm/softlayer_vm.go [hunk 2]
```diff
@@ -135,6 +150,21 @@ func (vm SoftLayerVM) AttachDisk(disk bslcdisk.Disk) error {
 		return bosherr.WrapErrorf(err, "Failed to attach volume with id %d to virtual guest with id: %d.", volume.Id, virtualGuest.Id)
 	}
 
+	metadata, err := bslcommon.GetUserMetadataOnVirtualGuest(vm.softLayerClient, virtualGuest.Id)
+
+	old_agentEnv, err := NewAgentEnvFromJSON(metadata)
+	new_agentEnv := old_agentEnv.AttachPersistentDisk(strconv.Itoa(disk.ID()), "/dev/sda")
```
Comment (maximilien) [inline softlayer/vm/softlayer_vm.go]: Are we sure the persistent disk name will always be: `/dev/sda` ?

File: softlayer/vm/softlayer_vm.go [hunk 3]
```diff
@@ -135,6 +150,21 @@ func (vm SoftLayerVM) AttachDisk(disk bslcdisk.Disk) error {
 		return bosherr.WrapErrorf(err, "Failed to attach volume with id %d to virtual guest with id: %d.", volume.Id, virtualGuest.Id)
 	}
 
+	metadata, err := bslcommon.GetUserMetadataOnVirtualGuest(vm.softLayerClient, virtualGuest.Id)
```
Comment (maximilien) [inline softlayer/vm/softlayer_vm.go]: You need to check `if err != nil` and return error here

File: softlayer/disk/softlayer_creator.go [hunk 4]
```diff
@@ -50,6 +50,7 @@ func (c SoftLayerCreator) Create(size int, cloudProps DiskCloudProperties, virtu
 }
 
 func (c SoftLayerCreator) getSoftLayerDiskSize(size int) int {
+	// Sizes and IOPS ranges: http://knowledgelayer.softlayer.com/learning/performance-storage-concepts
```
Comment (maximilien) [inline softlayer/disk/softlayer_creator.go]: Nice comment :smile:

**Comment by maximilien** [comment]
Hi, @jianqiu,

This is looking better. I have a few comments. Please address and I will merge.

I am not closing the PR so you can reuse it by adding new commits. When ready add a comment here to let me know.

Best,

@maximilien

File: softlayer/vm/softlayer_vm.go (hunk 2 above)
Comment (jianqiu) [inline softlayer/vm/softlayer_vm.go]: For the kind of Softlayer,  the first persistent iscsi disk would be `/dev/sda`. We will consider disk migration in next step.

**Comment by jianqiu** [comment]
Hi, @maximilien 

Thanks for your detailed comments, the fix had been included in the code, please review them, thanks!

File: softlayer/vm/softlayer_vm.go [hunk 5]
```diff
@@ -135,6 +150,21 @@ func (vm SoftLayerVM) AttachDisk(disk bslcdisk.Disk) error {
...
+
+	old_agentEnv, err := NewAgentEnvFromJSON(metadata)
```
Comment (maximilien) [inline softlayer/vm/softlayer_vm.go]: You did not check `if err != nil` here?

File: softlayer/vm/softlayer_vm.go (hunk 2 above)
Comment (maximilien) [inline softlayer/vm/softlayer_vm.go]: Please add a comment to say that this is always the case in SL and any links to documentation if you have.

**Comment by maximilien** [comment]
Getting very close. Two more issues: 1) you missed a check for `if err != nil` and 2) please add a comment about `/dev/sda` being fixed in SL.

Thanks.

**Comment by jianqiu** [comment]
@maximilien 

Thanks for your comments!  I am sorry for missing error handling again in my code, I should make it as a good habit.

**Comment by jianqiu** [comment]
@maximilien 

New changes had been included, please review them again, thanks!

**Comment by maximilien** [comment]
Ok cool. Merging. However, please ensure that there is an integration test that test this new code either directly or indirectly. If none, then add a chore to write this integration test. Thanks.

Cheers :beers:

**Pull request merged**